In [31]:
import numpy as np
import pandas as pd 
import matplotlib.pyplot as plt
import seaborn as sns

In [32]:
from sklearn.model_selection import train_test_split
from sklearn.datasets import load_wine
from sklearn.metrics import accuracy_score
from sklearn.naive_bayes import MultinomialNB,GaussianNB
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
from sklearn.pipeline import Pipeline

In [33]:
df = pd.read_csv('spam.csv')

In [34]:
df.head()

Category                                            Message
0      ham  Go until jurong point, crazy.. Available only ...
1      ham                      Ok lar... Joking wif u oni...
2     spam  Free entry in 2 a wkly comp to win FA Cup fina...
3      ham  U dun say so early hor... U c already then say...
4      ham  Nah I don't think he goes to usf, he lives aro...

In [35]:
df['Spam'] = pd.get_dummies(df['Category'],drop_first=True)
df

Category                                            Message  Spam
0         ham  Go until jurong point, crazy.. Available only ...     0
1         ham                      Ok lar... Joking wif u oni...     0
2        spam  Free entry in 2 a wkly comp to win FA Cup fina...     1
3         ham  U dun say so early hor... U c already then say...     0
4         ham  Nah I don't think he goes to usf, he lives aro...     0
...       ...                                                ...   ...
5567     spam  This is the 2nd time we have tried 2 contact u...     1
5568      ham               Will ü b going to esplanade fr home?     0
5569      ham  Pity, * was in mood for that. So...any other s...     0
5570      ham  The guy did some bitching but I acted like i'd...     0
5571      ham                         Rofl. Its true to its name     0

[5572 rows x 3 columns]

In [36]:
df = df.drop('Category',axis=1)

In [37]:
df

Message  Spam
0     Go until jurong point, crazy.. Available only ...     0
1                         Ok lar... Joking wif u oni...     0
2     Free entry in 2 a wkly comp to win FA Cup fina...     1
3     U dun say so early hor... U c already then say...     0
4     Nah I don't think he goes to usf, he lives aro...     0
...                                                 ...   ...
5567  This is the 2nd time we have tried 2 contact u...     1
5568               Will ü b going to esplanade fr home?     0
5569  Pity, * was in mood for that. So...any other s...     0
5570  The guy did some bitching but I acted like i'd...     0
5571                         Rofl. Its true to its name     0

[5572 rows x 2 columns]

In [38]:
x = df.drop('Spam',axis=1)
y=df['Spam']

In [39]:
y

0       0
1       0
2       1
3       0
4       0
       ..
5567    1
5568    0
5569    0
5570    0
5571    0
Name: Spam, Length: 5572, dtype: uint8

In [40]:
xtrain,xtest,ytrain,ytest = train_test_split(x,y,test_size=0.2,random_state=101)

In [41]:
xtrain

Message
1398  Then we wait 4 u lor... No need 2 feel bad lar...
1084  Wat makes some people dearer is not just de ha...
1086  FR'NDSHIP is like a needle of a clock. Though ...
2655                        Its sarcasm.. .nt scarcasim
1059  Tell me they're female :V how're you throwing ...
...                                                 ...
4171  Mmmmmm ... I love you,so much, Ahmad ... I can...
599   Will do. Was exhausted on train this morning. ...
1361  Yo dude guess who just got arrested the other day
1547              Shant disturb u anymore... Jia you...
4959                   Why didn't u call on your lunch?

[4457 rows x 1 columns]

In [42]:
cv = CountVectorizer()

In [43]:
xtrain_count = cv.fit_transform(xtrain['Message'])

In [44]:
xtrain_count.toarray()[:3]


array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]], dtype=int64)

In [45]:
xtrain_count

<4457x7774 sparse matrix of type '<class 'numpy.int64'>'
	with 59893 stored elements in Compressed Sparse Row format>

In [46]:
model = MultinomialNB()
model.fit(xtrain_count,ytrain)

MultinomialNB()

In [47]:
xtest_count = cv.transform(xtest['Message'])
mnbpred = model.predict(xtest_count)
accuracy_score(mnbpred,ytest)

0.9838565022421525

Therefore Naive Bayes Spam Classifier is **98.39%** accuracte when predicting spam emails

Using **tf-idf Vectorizer**

In [48]:
tfidf = TfidfVectorizer()
xtrain_tfidf = tfidf.fit_transform(xtrain['Message'])
xtest_tfidf  = tfidf.transform(xtest['Message'])

In [49]:
nbc = MultinomialNB()
nbc.fit(xtrain_tfidf,ytrain)
tfidfpreds = nbc.predict(xtest_tfidf)
accuracy_score(tfidfpreds,ytest)

0.9533632286995516

Therefore Naive Bayes Spam Classifier is **98.39%** accuracte when predicting spam emails using TFIDF Vectorizer

Using Pipelines

In [50]:
pipe = Pipeline([
    ('vectorizer',CountVectorizer()),
    ('nb',MultinomialNB())
])

In [51]:
#pipeline not working will check later

In [52]:
wines = load_wine()
wines

{'data': array([[1.423e+01, 1.710e+00, 2.430e+00, ..., 1.040e+00, 3.920e+00,
         1.065e+03],
        [1.320e+01, 1.780e+00, 2.140e+00, ..., 1.050e+00, 3.400e+00,
         1.050e+03],
        [1.316e+01, 2.360e+00, 2.670e+00, ..., 1.030e+00, 3.170e+00,
         1.185e+03],
        ...,
        [1.327e+01, 4.280e+00, 2.260e+00, ..., 5.900e-01, 1.560e+00,
         8.350e+02],
        [1.317e+01, 2.590e+00, 2.370e+00, ..., 6.000e-01, 1.620e+00,
         8.400e+02],
        [1.413e+01, 4.100e+00, 2.740e+00, ..., 6.100e-01, 1.600e+00,
         5.600e+02]]),
 'target': array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1,
        1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
        1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
        1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1

In [53]:
winedf = pd.DataFrame(wines.data,columns=wines.feature_names)

In [54]:
winedf

alcohol  malic_acid   ash  alcalinity_of_ash  magnesium  total_phenols  \
0      14.23        1.71  2.43               15.6      127.0           2.80   
1      13.20        1.78  2.14               11.2      100.0           2.65   
2      13.16        2.36  2.67               18.6      101.0           2.80   
3      14.37        1.95  2.50               16.8      113.0           3.85   
4      13.24        2.59  2.87               21.0      118.0           2.80   
..       ...         ...   ...                ...        ...            ...   
173    13.71        5.65  2.45               20.5       95.0           1.68   
174    13.40        3.91  2.48               23.0      102.0           1.80   
175    13.27        4.28  2.26               20.0      120.0           1.59   
176    13.17        2.59  2.37               20.0      120.0           1.65   
177    14.13        4.10  2.74               24.5       96.0           2.05   

     flavanoids  nonflavanoid_phenols  proanthocyanins  color_intensity   hue  \
0          3.06                  0.28             2.29             5.64  1.04   
1          2.76                  0.26             1.28             4.38  1.05   
2          3.24                  0.30             2.81             5.68  1.03   
3          3.49                  0.24             2.18             7.80  0.86   
4          2.69                  0.39             1.82             4.32  1.04   
..          ...                   ...              ...              ...   ...   
173        0.61                  0.52             1.06             7.70  0.64   
174        0.75                  0.43             1.41             7.30  0.70   
175        0.69                  0.43             1.35            10.20  0.59   
176        0.68                  0.53             1.46             9.30  0.60   
177        0.76                  0.56             1.35             9.20  0.61   

     od280/od315_of_diluted_wines  proline  
0                            3.92   1065.0  
1                            3.40   1050.0  
2                            3.17   1185.0  
3                            3.45   1480.0  
4                            2.93    735.0  
..                            ...      ...  
173                          1.74    740.0  
174                          1.56    750.0  
175                          1.56    835.0  
176                          1.62    840.0  
177                          1.60    560.0  

[178 rows x 13 columns]

In [55]:
winedf['target'] = pd.DataFrame(wines.target)
winedf

alcohol  malic_acid   ash  alcalinity_of_ash  magnesium  total_phenols  \
0      14.23        1.71  2.43               15.6      127.0           2.80   
1      13.20        1.78  2.14               11.2      100.0           2.65   
2      13.16        2.36  2.67               18.6      101.0           2.80   
3      14.37        1.95  2.50               16.8      113.0           3.85   
4      13.24        2.59  2.87               21.0      118.0           2.80   
..       ...         ...   ...                ...        ...            ...   
173    13.71        5.65  2.45               20.5       95.0           1.68   
174    13.40        3.91  2.48               23.0      102.0           1.80   
175    13.27        4.28  2.26               20.0      120.0           1.59   
176    13.17        2.59  2.37               20.0      120.0           1.65   
177    14.13        4.10  2.74               24.5       96.0           2.05   

     flavanoids  nonflavanoid_phenols  proanthocyanins  color_intensity   hue  \
0          3.06                  0.28             2.29             5.64  1.04   
1          2.76                  0.26             1.28             4.38  1.05   
2          3.24                  0.30             2.81             5.68  1.03   
3          3.49                  0.24             2.18             7.80  0.86   
4          2.69                  0.39             1.82             4.32  1.04   
..          ...                   ...              ...              ...   ...   
173        0.61                  0.52             1.06             7.70  0.64   
174        0.75                  0.43             1.41             7.30  0.70   
175        0.69                  0.43             1.35            10.20  0.59   
176        0.68                  0.53             1.46             9.30  0.60   
177        0.76                  0.56             1.35             9.20  0.61   

     od280/od315_of_diluted_wines  proline  target  
0                            3.92   1065.0       0  
1                            3.40   1050.0       0  
2                            3.17   1185.0       0  
3                            3.45   1480.0       0  
4                            2.93    735.0       0  
..                            ...      ...     ...  
173                          1.74    740.0       2  
174                          1.56    750.0       2  
175                          1.56    835.0       2  
176                          1.62    840.0       2  
177                          1.60    560.0       2  

[178 rows x 14 columns]

In [56]:
wines.target_names

array(['class_0', 'class_1', 'class_2'], dtype='<U7')

In [57]:
wxtrain,wxtest,wytrain,wytest = train_test_split(winedf.drop('target',axis=1),winedf['target'],test_size=0.2,random_state=101)

In [58]:
winetester = MultinomialNB()
winetester.fit(wxtrain,wytrain)
winepred = winetester.predict(wxtest)
accuracy_score(winepred,wytest)

0.7777777777777778

In [59]:
winetester = GaussianNB()
winetester.fit(wxtrain,wytrain)
winepred = winetester.predict(wxtest)
accuracy_score(winepred,wytest)

0.9722222222222222